README
INSTALLATION

  Installation is simple.  Follow the following steps:

	$ ruby extconf.rb
	$ make
	# make install

  If extconf yacks up an error, follow the instructions it provides.
  Once installed, look at the test scripts (*.rb, not extconf.rb) and
  make sure you run document_self.rb to see what's available.



DEPENDENCIES

  libxml requires a few other libraries to be installed inorder to
  function properly.

  *) libm      (math routines: very standard)
  *) libz      (zlib)
  *) libiconv
  *) libxml2


USAGE

  Basic usage for reading and writing documents.

  WRITING

  Writing a simple document:

    require 'xml/libxml'
    doc = XML::Document.new()
    doc.root = XML::Node.new('root_node')
    root = doc.root

    root << elem1 = XML::Node.new('elem1')
    elem1['attr1'] = 'val1'
    elem1['attr2'] = 'val2'

    root << elem2 = XML::Node.new('elem2')
    elem2['attr1'] = 'val1'
    elem2['attr2'] = 'val2'

    root << elem3 = XML::Node.new('elem3')
    elem3 << elem4 = XML::Node.new('elem4')
    elem3 << elem5 = XML::Node.new('elem5')

    elem5 << elem6 = XML::Node.new('elem6')
    elem6 << 'Content for element 6'

    elem3['attr'] = 'baz'

    # Namespace hack to reduce the numer of times XML:: is typed
    include XML
    root << elem7 = Node.new('foo')
    1.upto(10) do |i|
      elem7 << n = Node.new('bar')
      n << i
    end

    format = true
    doc.save('output.xml', format)

  The file output.xml contains:

    <?xml version="1.0"?>
    <root_node>
      <elem1 attr1="val1" attr2="val2"/>
      <elem2 attr1="val1" attr2="val2"/>
      <elem3 attr="baz">
        <elem4/>
        <elem5>
          <elem6>Content for element 6</elem6>
        </elem5>
      </elem3>
      <foo>
        <bar>1</bar>
        <bar>2</bar>
        <bar>3</bar>
        <bar>4</bar>
        <bar>5</bar>
        <bar>6</bar>
        <bar>7</bar>
        <bar>8</bar>
        <bar>9</bar>
        <bar>10</bar>
      </foo>
    </root_node>


  READING

  Reading XML is slightly more complex and there are many more ways to
  perform this operation.  This reads in and processes the above
  generated XML document, output.xml. This script assumes that the
  structure of the document is already known.

    require 'xml/libxml'
    doc = XML::Document.file('output.xml')
    root = doc.root

    puts "Root element name: #{root.name}"

    elem3 = root.find('elem3').to_a.first
    puts "Elem3: #{elem3['attr']}"

    doc.find('//root_node/foo/bar').each do |node|
      puts "Node path: #{node.path} \t Contents: #{node}"
    end

  And your terminal should look like:

    Root element name: root_node
    Elem3: baz
    Node path: /root_node/foo/bar[1]         Contents: 1
    Node path: /root_node/foo/bar[2]         Contents: 2
    Node path: /root_node/foo/bar[3]         Contents: 3
    Node path: /root_node/foo/bar[4]         Contents: 4
    Node path: /root_node/foo/bar[5]         Contents: 5
    Node path: /root_node/foo/bar[6]         Contents: 6
    Node path: /root_node/foo/bar[7]         Contents: 7
    Node path: /root_node/foo/bar[8]         Contents: 8
    Node path: /root_node/foo/bar[9]         Contents: 9
    Node path: /root_node/foo/bar[10]        Contents: 10


If you have any questions, please send email to devel@rubynet.org.

# $Id: README,v 1.1.1.1 2003/08/23 15:08:42 xpovolny Exp $


Changelog
14.4.2004 Mangler Jurgen <et@wkv.at>
    * ruby_xml_node.cz: fixed ruby_xml_node_property_set. The ill-behaviour
                was, that there was added a second attribute of the same
                name, when you were setting the value of an already existing
                attribute.

17.3.2004 Lukas Svoboda <luks@fi.muni.cz>
    * ruby_xml_node.c: ruby_xml_node_to_s now returns XML subtree dump.

27.2.2004 Martin Povolny <martin@solnet.cz>
    * ruby_xml_node.c: added XML::Node.copy, this makes possible building
                of xml documents from nodes taken from other xml documents
                without making ruby SIGSEGV (see tests/copy_bug.rb).

26.2.2004 Martin Povolny <martin@solnet.cz>
    * ruby_xml_dtd.c, ruby_xml_dtd.h, ruby_xml_schema.c, ruby_xml_schema.h:
                more work on validation, now you can actually validate
                document using dtd or xml schema, also solved warning and
                error propagation (see tests/{dtd|schema}-test.rb).

30.12.2003 Martin Povolny <martin@solnet.cz>
    * ruby_xml_dtd.c, ruby_xml_dtd.h, ruby_xml_schema.c, ruby_xml_schema.h:
                prelimitary support for dtd and schema validation

15.9.2003 Martin Povolny <martin@solnet.cz>
    * ruby_xml_input_cbg.c, libxml.c: added class InputCallbacks to make
                possible registering custom input callbacks
                handlers (xmlRegisterInputCallbacks) written in ruby

1.8.2003 Martin Povolny <martin@solnet.cz>
    * ruby_xml_document.c: corrected argument handling in ruby_xml_document_find
    * ruby_xml_node.c: corrected argument handling in ruby_xml_node_find


ruby_xml_attr.c
/* $Id: ruby_xml_attr.c,v 1.1.1.1 2003/08/23 15:08:42 xpovolny Exp $ */

/* Please see the LICENSE file for copyright and distribution information */

#include "libxml.h"
#include "ruby_xml_attr.h"

VALUE cXMLAttr;

void
ruby_xml_attr_free(ruby_xml_attr *rxa) {
  if (rxa->attr != NULL && !rxa->is_ptr) {
    xmlUnlinkNode((xmlNodePtr)rxa->attr);
    xmlFreeNode((xmlNodePtr)rxa->attr);
    rxa->attr = NULL;
  }

  free(rxa);
}


VALUE
ruby_xml_attr_child_get(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->children == NULL)
    return(Qnil);
  else
    return(ruby_xml_node_new2(cXMLNode, rxa->xd, rxa->attr->children));
}


VALUE
ruby_xml_attr_child_q(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->children == NULL)
    return(Qfalse);
  else
    return(Qtrue);
}


VALUE
ruby_xml_attr_doc_get(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->doc == NULL)
    return(Qnil);
  else
    return(ruby_xml_document_new(cXMLDocument, rxa->attr->doc));
}


VALUE
ruby_xml_attr_doc_q(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->doc == NULL)
    return(Qfalse);
  else
    return(Qtrue);
}


VALUE
ruby_xml_attr_last_get(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->last == NULL)
    return(Qnil);
  else
    return(ruby_xml_node_new2(cXMLNode, rxa->xd, rxa->attr->last));
}


VALUE
ruby_xml_attr_last_q(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->last == NULL)
    return(Qfalse);
  else
    return(Qtrue);
}


static void
ruby_xml_attr_mark(ruby_xml_attr *rxa) {
  if (rxa == NULL) return;
  if (!NIL_P(rxa->xd)) rb_gc_mark(rxa->xd);
}


VALUE
ruby_xml_attr_name_get(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);

  if (rxa->attr->name == NULL)
    return(Qnil);
  else
    return(rb_str_new2(rxa->attr->name));
}


VALUE
ruby_xml_attr_new(VALUE class, VALUE xd, xmlAttrPtr attr) {
  ruby_xml_attr *rxa;

  rxa = ALLOC(ruby_xml_attr);
  rxa->attr = attr;
  rxa->xd = xd;
  rxa->is_ptr = 0;
  return(Data_Wrap_Struct(class, ruby_xml_attr_mark,
			  ruby_xml_attr_free, rxa));
}


VALUE
ruby_xml_attr_new2(VALUE class, VALUE xd, xmlAttrPtr attr) {
  ruby_xml_attr *rxa;

  rxa = ALLOC(ruby_xml_attr);
  rxa->attr = attr;
  rxa->xd = xd;
  rxa->is_ptr = 1;
  return(Data_Wrap_Struct(class, ruby_xml_attr_mark,
			  ruby_xml_attr_free, rxa));
}


VALUE
ruby_xml_attr_next_get(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->next == NULL)
    return(Qnil);
  else
    return(ruby_xml_attr_new(cXMLAttr, rxa->xd, rxa->attr->next));
}


VALUE
ruby_xml_attr_next_q(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->next == NULL)
    return(Qfalse);
  else
    return(Qtrue);
}


/* I think libxml2's naming convention blows monkey ass */
VALUE
ruby_xml_attr_node_type_name(VALUE self) {
  return(rb_str_new2("attribute"));
}


VALUE
ruby_xml_attr_ns_get(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->ns == NULL)
    return(Qnil);
  else
    return(ruby_xml_ns_new2(cXMLNS, rxa->xd, rxa->attr->ns));
}


VALUE
ruby_xml_attr_ns_q(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->ns == NULL)
    return(Qfalse);
  else
    return(Qtrue);
}


VALUE
ruby_xml_attr_parent_get(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->parent == NULL)
    return(Qnil);
  else
    return(ruby_xml_node_new2(cXMLNode, rxa->xd, rxa->attr->parent));
}


VALUE
ruby_xml_attr_parent_q(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->parent == NULL)
    return(Qfalse);
  else
    return(Qtrue);
}


VALUE
ruby_xml_attr_prev_get(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->prev == NULL)
    return(Qnil);
  else
    return(ruby_xml_attr_new(cXMLAttr, rxa->xd, rxa->attr->prev));
}


VALUE
ruby_xml_attr_prev_q(VALUE self) {
  ruby_xml_attr *rxa;
  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (rxa->attr->prev == NULL)
    return(Qfalse);
  else
    return(Qtrue);
}


VALUE
ruby_xml_attr_value(VALUE self) {
  ruby_xml_attr *rxa;
  char *value;

  Data_Get_Struct(self, ruby_xml_attr, rxa);
  if (ruby_xml_attr_parent_q(self) == Qtrue) {
    value = xmlGetProp(rxa->attr->parent, rxa->attr->name);
    if (value != NULL)
      return(rb_str_new2(value));
  }
  return(Qnil);
}


void
ruby_init_xml_attr(void) {
  cXMLAttr = rb_define_class_under(mXML, "Attr", rb_cObject);
  rb_define_method(cXMLAttr, "child", ruby_xml_attr_child_get, 0);
  rb_define_method(cXMLAttr, "child?", ruby_xml_attr_child_q, 0);
  rb_define_method(cXMLAttr, "doc", ruby_xml_attr_doc_get, 0);
  rb_define_method(cXMLAttr, "doc?", ruby_xml_attr_doc_q, 0);
  rb_define_method(cXMLAttr, "last", ruby_xml_attr_last_get, 0);
  rb_define_method(cXMLAttr, "last?", ruby_xml_attr_last_q, 0);
  rb_define_method(cXMLAttr, "name", ruby_xml_attr_name_get, 0);
  rb_define_method(cXMLAttr, "next", ruby_xml_attr_next_get, 0);
  rb_define_method(cXMLAttr, "next?", ruby_xml_attr_next_q, 0);
  rb_define_method(cXMLAttr, "node_type_name", ruby_xml_attr_node_type_name, 0);
  rb_define_method(cXMLAttr, "ns", ruby_xml_attr_ns_get, 0);
  rb_define_method(cXMLAttr, "ns?", ruby_xml_attr_ns_q, 0);
  rb_define_method(cXMLAttr, "parent", ruby_xml_attr_parent_get, 0);
  rb_define_method(cXMLAttr, "parent?", ruby_xml_attr_parent_q, 0);
  rb_define_method(cXMLAttr, "prev", ruby_xml_attr_prev_get, 0);
  rb_define_method(cXMLAttr, "prev?", ruby_xml_attr_prev_q, 0);
  rb_define_method(cXMLAttr, "value", ruby_xml_attr_value, 0);
}
